components/Ripple/src/MDCStatefulRippleView.m

```diff
@@ -0,0 +1,229 @@
+    return;
+  }
+  _selected = selected;
+  // Go into the selected state visually.
+  if (selected) {
+    if (!self.activeRippleLayer) {
+      // If we go into the selected state manually, without coming from the highlighted state,
+      // We present the ripple overlay instantly without animation.
+      [self updateRippleColor];
+      [self beginRippleTouchDownAtPoint:_lastTouch animated:NO completion:nil];
+    } else {
+      [self updateActiveRippleColor];
+    }
+  } else {
+    // If we are no longer selecting, we cancel all the ripples.
+    [self updateRippleColor];
+    [self cancelAllRipplesAnimated:YES completion:nil];
+  }
+}
+
+- (void)setRippleHighlighted:(BOOL)rippleHighlighted {
+  if (rippleHighlighted == _rippleHighlighted) {
+    return;
+  }
+  _rippleHighlighted = rippleHighlighted;
+  // Go into the highlighted state visually.
+  if (rippleHighlighted && !_tapWentInsideOfBounds) {
+    // If ripple becomes highlighted we initiate a ripple with animation.
+    [self updateRippleColor];
+    [self beginRippleTouchDownAtPoint:_lastTouch animated:_didReceiveTouch completion:nil];
+  } else {
+    if (!self.allowsSelection && !self.selected && !self.dragged && !_tapWentOutsideOfBounds) {
+      // We dissolve the ripple when highlighted is NO, unless we are going into selection.
+      [self updateRippleColor];
+      [self beginRippleTouchUpAnimated:YES completion:nil];
+    }
+  }
+}
+
+- (void)setDragged:(BOOL)dragged {
+  if (dragged == _dragged) {
+    return;
+  }
+  _dragged = dragged;
+  // Go into the dragged state visually.
+  if (dragged) {
+    if (!self.activeRippleLayer) {
+      // If we go into the dragged state manually, without coming from the highlighted state,
+      // We present the ripple overlay instantly without animation.
+      [self updateRippleColor];
+      [self beginRippleTouchDownAtPoint:_lastTouch animated:NO completion:nil];
+    } else {
+      [self updateActiveRippleColor];
+    }
+  } else {
+    // If we are no longer dragging, we cancel all the ripples.
+    [self updateRippleColor];
+    [self cancelAllRipplesAnimated:YES completion:nil];
+  }
+}
+
+- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
+  UITouch *touch = [touches anyObject];
+  CGPoint location = [touch locationInView:self];
+  _didReceiveTouch = YES;
+  _lastTouch = location;
+  _tapWentInsideOfBounds = NO;
+  _tapWentOutsideOfBounds = NO;
+  [super touchesBegan:touches withEvent:event];
+}
+
+- (void)touchesMoved:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
+  // When the touch is held and moved outside and inside the bounds of the surface,
+  // the ripple should gracefully fade out and in accordingly.
+  UITouch *touch = [touches anyObject];
+  CGPoint location = [touch locationInView:self];
+  BOOL pointContainedinBounds = CGRectContainsPoint(self.bounds, location);
+  if (pointContainedinBounds && _tapWentOutsideOfBounds) {
+    if (_tapWentOutsideOfBounds) {
```

Comment:
This is a redundant conditional with the line above: you can delete it.

Reply:
Done, thanks.